### ISSUE TITLE
panic: runtime error: index out of range

### ISSUE THREAD

**Issue by recollir** [ExpressenAB/bigip_exporter#5]
Running the latest commit  b415f1f3502fe7c2d7cb7732698dec57b26f4834 I get the following runtime error.

This is using version "BIG-IP 12.1.1 Build 1.0.196 Hotfix HF1". The runtime error occurs as soon as I hit the /metrics endpoint.

$ ./bigip_exporter -bigip.host bigip.example.com -bigip.port 443 -bigip.username admin -bigip.password admin
2016/11/16 15:40:11 VS was succes: true
2016/11/16 15:40:11 Getting VS stats took 394.16887ms
2016/11/16 15:40:11 Pool was succes: true
2016/11/16 15:40:11 Getting pool stats took 553.608348ms
2016/11/16 15:40:11 Node was succes: true
2016/11/16 15:40:11 Getting node stats took 706.493054ms
panic: runtime error: index out of range

goroutine 9 [running]:
panic(0x2da280, 0xc420012110)
        /usr/local/Cellar/go/1.7/libexec/src/runtime/panic.go:500 +0x1a1
github.com/ExpressenAB/bigip_exporter/collector.(*ruleCollector).Collect(0xc4200f44c0, 0xc42018e2a0)
        /Users/federico/projects/golang/src/github.com/ExpressenAB/bigip_exporter/collector/rule_collector.go:153 +0x8db
github.com/ExpressenAB/bigip_exporter/collector.(*bigipCollector).Collect.func1(0xc42018e2a0, 0xc420013d40, 0x48f9a0, 0xc4200f44c0)
        /Users/federico/projects/golang/src/github.com/ExpressenAB/bigip_exporter/collector/bigip_collector.go:44 +0x3b
created by github.com/ExpressenAB/bigip_exporter/collector.(*bigipCollector).Collect
        /Users/federico/projects/golang/src/github.com/ExpressenAB/bigip_exporter/collector/bigip_collector.go:46 +0x141

**Comment by cgunning** [comment]
Hi. Thanks for your issue. I think I know where the error arises but since I don't have access to a BIG-IP of that version I have difficulties testing my hypothesis and how to solve it. I make certain assumptions about the response I get from BIG-IP. You wouldn't be able to make a manual API call to the `/mgmt/tm/ltm/rule/stats` endpoint and post the structure of the response? In my case it looks for example like this:

``` json
"https://<host>/mgmt/tm/ltm/rule/~<partition>~<rule_name>:HTTP_RESPONSE/stats": {
      "nestedStats": {
        "kind": "tm:ltm:rule:rulestats",
        "selfLink": "https://<host>/mgmt/tm/ltm/rule/~<partition~<rule_name>:HTTP_RESPONSE/stats?ver=12.0.0",
        "entries": {
          "aborts": {
            "value": 0
          },
          "avgCycles": {
            "value": 0
          },
          "eventType": {
            "description": "HTTP_RESPONSE"
          },
          "failures": {
            "value": 0
          },
          "maxCycles": {
            "value": 0
          },
          "minCycles": {
            "value": 0
          },
          "tmName": {
            "description": "<description>"
          },
          "priority": {
            "value": 100
          },
          "totalExecutions": {
            "value": 0
          }
        }
      }
```

The part I'm especially interested in is the first key that in my case looks lite this: `https://<host>/mgmt/tm/ltm/rule/~<partition>~<rule_name>:HTTP_RESPONSE/stats`

**Comment by recollir** [comment]
Yes. I assume the same. Something has changed in the JSON returned. And yesterday late night I then got the idea of running the exporter queries and the responses it gets through Charlesproxy to record everything and save it. But I can also do your suggestions manually. Unfortunately I will not be able to do this before late today or tomorrow morning.

**Comment by cgunning** [comment]
I'm just grateful if you're able to help me find the issue. The problem I think lies in the following lines of code (line 147-153 in `collector/rule_collector.go`). This parses the `~<partition>~<rule_name>:HTTP_RESPONSE` part into it's separate fields.

```
keyParts := strings.Split(key, "/")
path := keyParts[len(keyParts)-2]
pathParts := strings.Split(path, "~")
partition := pathParts[1]
eventParts := strings.Split(pathParts[2], ":")
ruleName := eventParts[0]
event := eventParts[1]
```

This is not done in the smartest of ways so I'll make an effort to do it a bit smarter as soon as I know your case

**Comment by recollir** [comment]
Ok, I did run the query against the /mgmt/tm/ltm/rule/stats endpoint. And extracted the part you are interested in - I will only post that one for one element as I then don't have to obfuscate the entire output. ;-)

Our bigip returns as well <partion> and <rule-name>. Also with ~. What I see is that our rule names not only contain [a-zA-Z0-9] but also characters like . - _ 

Example: 

"https://localhost/mgmt/tm/ltm/rule/~Common~xxx.yyy.zzz-pool-selector:HTTP_REQUEST/stats" or
"https://localhost/mgmt/tm/ltm/rule/~Common~aaa-bbb-80-rewrite-irule:HTTP_RESPONSE/stats"

I post one of the elements - all other look the same - I checked the json object with jq and the elements are consistent. 

```
"https://localhost/mgmt/tm/ltm/rule/~Common~www-xxxxxx-access-irule:HTTP_REQUEST/stats": {
  "nestedStats": {
    "kind": "tm:ltm:rule:rulestats",
    "selfLink": "https://localhost/mgmt/tm/ltm/rule/~Common~www-xxxxxx-access-irule:HTTP_REQUEST/stats?ver=12.1.1",
    "entries": {
      "aborts": {
        "value": 0
      },
      "avgCycles": {
        "value": 8460
      },
      "eventType": {
        "description": "HTTP_REQUEST"
      },
      "failures": {
        "value": 0
      },
      "maxCycles": {
        "value": 6610270
      },
      "minCycles": {
        "value": 2955
      },
      "tmName": {
        "description": "/Common/www-xxxxxx-access-irule"
      },
      "priority": {
        "value": 500
      },
      "totalExecutions": {
        "value": 5751229
      }
    }
  }
}, 
```

**Comment by cgunning** [comment]
Hmm. Do you have the possibility to add a bit of logging in order to see exactly what it is that is breaking. I'll work on adding more logging on my end but in order to pinpoint the problem as fast as possible it would be great if you could test doing that.

**Comment by recollir** [comment]
I am already adding fmt.Println statements in the code block you mentioned. :-)

**Comment by cgunning** [comment]
Great! Thanks :)

**Comment by recollir** [comment]
I found it. The terminal output of jq for all 100+ keys hided the fact that the naming convention can include three ~ in the keyname. It seems that this happens if the bigip is "configured" through an iApp:

`https://localhost/mgmt/tm/ltm/rule/~<partition>~<iapp-name>~<rule-name>:ACCESS_ACL_ALLOWED/stats`

We have 11 of these in the 100+ output of the stats endpoint.

The following split with ':' will then fail of course.

I guess a quick fix would be to not use pathParts[2] but to choose the last element of that array. You get partition and rule-name but not the app-name. That should be OK.

/F

**Comment by recollir** [comment]
It works. I did the following and you will get a pull request...

```
-                       eventParts := strings.Split(pathParts[2], ":")
+                       eventParts := strings.Split(pathParts[len(pathParts)-1], ":")
```

/F

**Comment by cgunning** [comment]
Solved in #6
